libgpopt/src/base/CUtils.cpp

```
@@ -589,37 +605,37 @@ CUtils::PexprScalarCmp
 	CExpression *pexprNewRight = pexprRight;
 
 	IMDId *pmdidCmpOp = NULL;
+
 	if (CMDAccessorUtils::FCmpExists(pmda, pmdidLeft, pmdidRight, ecmpt))
 	{
 		pmdidCmpOp = PmdidScCmp(pmda, pmdidLeft, pmdidRight, ecmpt);
 	}
-	else
+	else if (CMDAccessorUtils::FCmpExists(pmda, pmdidLeft, pmdidLeft, ecmpt) && CMDAccessorUtils::FCastExists(pmda, pmdidRight, pmdidLeft))
 	{
-		// generate a scalar comparison by casting one of the two sides
-		if (CMDAccessorUtils::FCastExists(pmda, pmdidLeft, pmdidRight))
-		{
-			pexprNewLeft = PexprCast(pmp, pmda, pexprLeft, pmdidRight);
-			pmdidCmpOp = PmdidScCmp(pmda, pmdidRight, pmdidRight, ecmpt);
-		}
-		else
-		{
-			GPOS_ASSERT(CMDAccessorUtils::FCastExists(pmda, pmdidRight, pmdidLeft));
-			pexprNewRight = PexprCast(pmp, pmda, pexprRight, pmdidLeft);
-			pmdidCmpOp = PmdidScCmp(pmda, pmdidLeft, pmdidLeft, ecmpt);
-		}
+		pexprNewRight = PexprCast(pmp, pmda, pexprRight, pmdidLeft);
+		pmdidCmpOp = PmdidScCmp(pmda, pmdidLeft, pmdidLeft, ecmpt);
+	}
+	else if (CMDAccessorUtils::FCmpExists(pmda, pmdidRight, pmdidRight, ecmpt) && CMDAccessorUtils::FCastExists(pmda, pmdidLeft, pmdidRight))
+	{
+		pexprNewLeft = PexprCast(pmp, pmda, pexprLeft, pmdidRight);
+		pmdidCmpOp = PmdidScCmp(pmda, pmdidRight, pmdidRight, ecmpt);
 	}
```

We need to raise an exception here. This is wrong!